cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,327 @@
+        // We would need to check both ways in swap cases that original brokerSet of both replicas matches the brokerSet of destination broker
+        if (action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP
+            && doesReplicaMoveViolateActionAcceptance(clusterModel,
+                                                      clusterModel.broker(action.destinationBrokerId())
+                                                                  .replica(action.destinationTopicPartition()),
+                                                      clusterModel.broker(action.sourceBrokerId()))) {
+          return REPLICA_REJECT;
+        }
+        return ACCEPT;
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  protected boolean doesReplicaMoveViolateActionAcceptance(ClusterModel clusterModel, Replica sourceReplica, Broker destinationBroker) {
+    // Destination broker cannot be in any other brokerSet.
+    // A replica's current broker set id can be different from the expected broker id defined by replica broker set mapping policy
+    try {
+      String expectedBrokerSetId =
+          _replicaToBrokerSetMappingPolicy.brokerSetIdForReplica(sourceReplica, clusterModel, _brokerSetResolutionHelper);
+      String destinationBrokerSetId = _brokerSetResolutionHelper.brokerSetId(destinationBroker);
+
+      return !expectedBrokerSetId.equals(destinationBrokerSetId);
+    } catch (BrokerSetResolutionException | ReplicaToBrokerSetMappingException e) {
+      LOG.error(String.format("[%s] Replica move violated action acceptance", name()), e);
+      return true;
+    }
+  }
+
+  @Override
+  public ClusterModelStatsComparator clusterModelStatsComparator() {
+    // Stats are irrelevant to a hard goal. The optimization would already fail if the goal requirements are not met.
+    return new GoalUtils.HardGoalStatsComparator();
+  }
+
+  @Override
+  public ModelCompletenessRequirements clusterModelCompletenessRequirements() {
+    // BrokerSet-awareness does not need load information at all
+    return new ModelCompletenessRequirements(MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING, 0.0, true);
+  }
+
+  /**
+   * While attempting to apply balancing action during re-balance,
+   * check if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * {@code false} otherwise.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param action       Action containing information about potential modification to the given cluster model.
+   * @return {@code true} if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * {@code false} otherwise.
+   */
+  @Override
+  protected boolean selfSatisfied(ClusterModel clusterModel, BalancingAction action) {
+    return actionAcceptance(action, clusterModel) == ACCEPT;
```

Comment:
Do we need to call `actionAcceptance(action, clusterModel)` here? -- i.e. aren't we sure that for this goal, whenever this function is called, we already know that we will get a `true` as the response?

If that is the case, can we just return `true` on this function?

Reply:
We are sure, but i wasn't fully aware of the usage of self satisfied function. Removing.